modules/gapi/src/executor/gapi_tbb_executor.cpp

```diff
@@ -0,0 +1,362 @@
+                while(!q.empty()){
+
+                    auto push_ready_dependees = [&q](tile_node* node) -> std::size_t
+                    {
+                        ITT_AUTO_TRACE_GUARD(ittTbbAddReadyBlocksToQueue);
+                        std::size_t ready_items = 0;
+                        //then enable task dependees
+                        for (auto* dependee : node->dependees)
+                        {
+                            //fetch_and_sub returns previous value
+                            if (1 == dependee->dependency_count.fetch_sub(1))
+                            {
+                                //tile node is ready for execution, add it to the queue
+                                q.push(dependee);
+                                ++ready_items;
+                            }
+                        }
+                        return ready_items;
+                    };
+
+                    tile_node* node = nullptr;
+                    bool poped = q.try_pop(node);
+                    ASSERT(poped && "queue should be non empty as we push items to it before we spawn");
+
+                    auto result = parallel::use_tbb_scheduler_bypass::no;
+                    //execute the task
+                    if (!node->async)
+                    {
+                        node->task_body();
+
+                        std::size_t ready_items = push_ready_dependees(node);
+
+                        if (ready_items > 0){
+                            //spawn one less tasks and say TBB to reuse(recycle) current task
+                            parallel::batch_spawn(ready_items - 1,root.get(), body);
+                            result = parallel::use_tbb_scheduler_bypass::yes;
+                        }
+                    }
+                    else
+                    {
+                        LOG_DEBUG(NULL, "Async task ");
+                        //move this lock (through copy) into the callback to block master until async tasks complete
+                        master_thread_sleep_lock_t block_master{&async_tasks};
+
+                        auto callback = [push_ready_dependees, node, &arena, &root, block_master, body] () mutable /*due to block_master.unlock()*/
+                        {
+                            LOG_DEBUG(NULL, "Async task callback is called ");
+                            std::size_t ready_items = push_ready_dependees(node);
+                            if (ready_items > 0)
+                            {
+                                //force master thread (one that do wait_for_all()) to (actively) wait for enqueued tasks
+                                //and unlock it right after all dependee tasks are spawn (and therefore ref_count of root has been increased accordingly)
+                                auto new_root_ref_count = root->add_ref_count(1);
+                                unlock_root_wait_t unlock_root_wait_guard{root.get()};
```

Comment:
Formally we can - and it will work. How ether there is a problem here : `wait_for_all` call will not return control until `ref_count` drops to zero. In practice this means that in the absence of TBB work (i.e. when all active tasks are async, i.e. GPU ones)  TBB scheduler  will search for tasks to execute infinitely (assuming they are present as the root `ref_count` value suggests), causing high CPU load (in the absence of CPU work :))